logic/Materialiser.java

```diff
@@ -80,20 +80,20 @@ private static Optional<Materialisation> materialise(Conclusion.Has.Explicit.Mat
     }
 
     private static Optional<Attribute> getAttribute(AttributeType attrType, ValueConstraint<?> value) {
-        if (attrType.isDateTime()) return Optional.ofNullable(attrType.asDateTime().get(value.asDateTime().value()));
-        else if (attrType.isBoolean()) return Optional.ofNullable(attrType.asBoolean().get(value.asBoolean().value()));
-        else if (attrType.isDouble()) return Optional.ofNullable(attrType.asDouble().get(value.asDouble().value()));
-        else if (attrType.isLong()) return Optional.ofNullable(attrType.asLong().get(value.asLong().value()));
-        else if (attrType.isString()) return Optional.ofNullable(attrType.asString().get(value.asString().value()));
+        if (attrType.isDateTime()) return Optional.ofNullable(attrType.asDateTime().get(value.asConstant().asDateTime().value()));
```

Comment:
we layer in a new `ValueConstraint` subtype, which is called `Constant` (sibling to `Variable`). The `Constant` now contains `Long/Double/Boolean/String/DateTime`